### ISSUE TITLE
After 3.8.x -> 4.13.x upgrade, connections to primary only increase

### ISSUE THREAD

**Issue by markstos** [Automattic/mongoose#6117]
**Do you want to request a *feature* or report a *bug*?**

Bug.

**What is the current behavior?**

Connection count to the primary only increase, until either the server restarts or the application restarts. This continues until the hardware resources are exhausted and a crash or restart is forced.

![Connection counts gone crazy after Mongoose 4.13 upgarde](https://lh3.googleusercontent.com/BveYc9xYzzbsdtaH15l0-4NR-78Ei3eKAqBehKFP8ZFELpoxPH3vDiGVNpRt0-Yem92aehnCi8dZdi1PTDomVge1plWPucwegJ4vE2ge4pnQ7LdW4RuOwmIXOhaYSGeJWgUqJOAQugZPZ0iD5Crkv23wvHANv3-KBV-MnPywD0T55Jjvke_meoDfzQMmVJnIzYCm4TbKFVtY2ao0yz0A6uc26DkR8UELXOikj_mDC_SOJPxD7udWwJiUUyyAU6wIp4TJ0_kj6qHDhcE-4bnirhPqyxgSIgBM-UNprR2u1Ok3sMFACCdXbldGLNx9iX291erNUUaBTmbVw8eqs9Xa3WqJEjL_RjDhdWsZr2rXhP1DZA7EE2kFVoDKwg9-xybSKZvTL95MINvOer0X0FGMaPiS0pCUTClecnNoKZNYbzRXdI7rPjmnISzcOPHWtY2m10fEObXq2mr1ZFpJfb7mQPM8OtVlFXUzn10AGLFF2zgnJ26EEOFk--4zWq-Gh_YbWam6d57CPd7e9aFV3GuBcgA_rUQmaUMihPcVTHiQGYxMckrE7cVsinE6z3zLwhv4Git7yI_8N8h4U57vVwaVdx82XWjoTkJpV-Zh5YGQQjZH_MQ8BKmF1EFVzdN3PrEpVmyWxeYhfACRAon0L2QkE4CHTsc-5cci=w1234-h385-no)

**If the current behavior is a bug, please provide the steps to reproduce.**

I don't a have a "minimal test case". 

We did start using `useMongoClient:true` as part of the upgrade. The docs only say that there is "new connection logic", but see now that the differences between the new and old logic are not detailed. Maybe this is related?

**What is the expected behavior?**

I expected the connection count to be stable as before. 

**Please mention your node.js, mongoose and MongoDB version.**

Node.js 4.8.7
Mongoose 4.13.8 (soon to upgrade to 4.13.11)
MongoDB 3.2.9

I realize this issue will be hard to act on without code to reproduce it, but knowing that the issue is a possibility will be of interest to other Mongoose users who haven't upgraded. When we get the issue resolved, I'll post the solution here. (If possible causes are not already known). 

It appears something is opening connections and never closing them, but I'm not sure what.

**Comment by TrejGun** [comment]
hmm interesting, i guess i can confirm this, my scheduled jobs are leaking a bit however i was not investigating

**Comment by vkarpov15** [comment]
@markstos can you show your connection logic with the sensitive credentials omitted?

**Comment by markstos** [comment]
We have a web app which connects to the database once at boot with a `mongoose.connect(...)`. Our connection options looked like this:

```
 options : {
      useMongoClient: true,
      reconnectTries: 10,
      reconnectInterval: 500,
      poolSize: 150,
},
```
What's more unique about situation is that we host about *120* copies of the web app configured that way, all pointed to the same MongoDB cluster. All the copies are due various customer configurations, plus redundancy across multiple servers. 

As we looked into this issue, we found the poolSize of 150 was set *much* higher than we actually needed. Today we tuned the poolSize down to `20` per instance. Since then, connections have stayed stable at around 2k on the primary. We also see a reduction in network I/O and disk I/O on the secondaries.

We are going to keep monitoring the cluster of coming days, but right-sizing the maximum size of our connection pool may be the change we needed.
